derivation/src/main/scala/caliban/derivation/DerivationMacros.scala

```diff
@@ -0,0 +1,240 @@
+package caliban.derivation
+
+import caliban.derivation.annotation.GQLExclude
+
+import scala.reflect.macros.blackbox
+
+/**
+ * Macro bundle for deriving instances of Schema for concrete classes,
```

Comment:
Originally so; following @ghostdogpr's comment, if I get this macro to do the same for sealed traits as magnolia currently does, this comment will be amended :-)

Reply:
There's a difference though: Magnolia strictly separates case classes (which become graphql `objects`) from sealed traits (which become graphql `enums`, `unions` or `interfaces`). But I think @Krever is looking for an actual trait like this:

```scala
trait MyBusinessLogic {
  def someMethod(a: A): B = ???
}

// somewhere else we create this trait by calling new MyBusinessLogic { }
```

This is not available in Magnolia at all. This kind of trait would be more similar to a case class: we would need to generate an `object` out of it.

Reply:
Yes, this is what I meant. I tend to separate my interfaces from actual implementation, mostly for the sake of readability and sometimes for testing. So I would love to be able to annotate and derive instance for a trait and use it later on for the actual implementation. 
IUUC the trait has all the (info) signatures needed for the derivation, right?
Also, I think sangria supports it, if that matters.

Reply:
I'm not sure if sangria supports it, but certainly in caliban's case it's not really possible to derive a Schema instance from the sealed trait _alone_; what would the `resolve` method return?

That said, if I'm successful in my effort to extend the macro to sealed traits and case objects (and I think I'm fairly close), you'd be able to define abstract methods on a trait, that its concrete subclasses will implement, which will translate into GraphQL fields/functions.

E.g.
```
sealed trait Foo {
  def method(x: A, b: B): C
}

case class Bar(arg1: D, arg2: E) extends Foo {
  def method(x: A, b: B): C = // actual implementation
}
```
should derive as expected.

Reply:
Perhaps I'm missing something. @Krever, could you provide an example of a trait you'd like to use as an interface, coupled with what a corresponding Schema definition would look like?

Reply:
If we take my example above, it would generate this schema:
```graphql
type MyBusinessLogic {
  someMethod(a: A!): B!
}
```
In short, the same thing as `case class MyBusinessLogic(someMethod: AArgs => B)`, except this time you don't need the `AArgs` wrapper class.

Reply:
What I meant was the derivation from any `trait`, not `sealed trait` specifically. This is a truncated example from one of my projects relying on sangria:
```
trait Mutations {

  @GraphQLField
  def approveQuestionnaire(
      questionnaireId: UUID,
      submission: Json,
  ): FutureValue[Context, SubmitResult]

}

trait Queries {

  @GraphQLField
  def getQuestionnaireData(
      ctx: Context[CrowContext, Unit],
      id: UUID,
  ): FutureValue[Context, QuestionnaireInsightResult]

}

case class Context(
    mutationService: Mutation,
    queryService: Query,
)

val MutationType =
  deriveContextObjectType[Context, Mutation, Unit](_.mutationService)
val QueryType =
  deriveContextObjectType[Context, Query, Unit](_.queryService)
val webformsSchema = Schema(QueryType, Some(MutationType))
```
I'm not the one who wrote it, so it might be the case that I'm missing something here. Anyway, this code derives a schema as @ghostdogpr, each parameter becomes an input and method name is operation name.

Reply:
If we're talking about deriving a GraphQL schema for arbitrary (non-sealed) traits, with abstract methods, the best we could do is statically derive a sort of proto-Schema (which has the type information, but no notion of how to actually resolve fields/functions on this type), supplemented with some runtime, dynamic functionality for registering specific implementations with it.

Reply:
@argson67 Coul you explain why? If the signatures are stable and dont change, regardless of the trait implementation, why can't we derive the final schema?

Reply:
Oh, I see. Thanks for the clarification. I'll take a poke at it, but it does indeed seem possible.

Reply:
Yup, I think I agree on the magical part, the annotation requirement sounds like good protection against undefined behavior.  And yes, it's totally orthogonal to the sealed trait. 
And I don't want to put any demands here, what you already did is great. I just wanted to share a usage pattern we have with sangria that was originally my motivation for raising #218